File: src/main/java/com/gitee/authentication/accounts/GiteeAccount.java
/*
 *  Copyright 2016-2019 码云 - Gitee
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.gitee.authentication.accounts;

import com.gitee.api.GiteeServerPath;
import com.intellij.util.xmlb.annotations.Attribute;
import com.intellij.util.xmlb.annotations.Property;
import com.intellij.util.xmlb.annotations.Tag;
import com.intellij.util.xmlb.annotations.Transient;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.UUID;

@Tag("account")
public class GiteeAccount {
	@Attribute("id")
	@NotNull
	private final String myId;

	@Attribute("name")
	@NotNull
	private String myName;

	@Property(style = Property.Style.ATTRIBUTE, surroundWithTag = false)
	@NotNull
	private final GiteeServerPath myServer;

	// serialization
	@SuppressWarnings("unused")
	private GiteeAccount() {
		myId = "";
		myName = "";
		myServer = new GiteeServerPath();
	}

	GiteeAccount(@NotNull String name, @NotNull GiteeServerPath server) {
		myId = UUID.randomUUID().toString();
		myName = name;
		myServer = server;
	}

	@Override
	public String toString() {
		return myServer + "/" + myName;
	}

	@NotNull
	String getId() {
		return myId;
	}

	@NotNull
	public String getName() {
		return myName;
	}

	@Transient
	public void setName(@NotNull String name) {
		myName = name;
	}

	@NotNull
	public GiteeServerPath getServer() {
		return myServer;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof GiteeAccount)) return false;
		GiteeAccount account = (GiteeAccount) o;
		return Objects.equals(myId, account.myId);
	}

	@Override
	public int hashCode() {
		return Objects.hash(myId);
	}
}

File: src/main/kotlin/com/gitee/authentication/accounts/GiteeAccount.kt
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.gitee.authentication.accounts

import com.gitee.api.GiteeServerPath
import com.intellij.collaboration.auth.ServerAccount
import com.intellij.openapi.util.NlsSafe
import com.intellij.util.xmlb.annotations.Attribute
import com.intellij.util.xmlb.annotations.Property
import com.intellij.util.xmlb.annotations.Tag
import com.intellij.util.xmlb.annotations.Transient
import org.jetbrains.annotations.VisibleForTesting

@Tag("account")
class GiteeAccount(
  @set:Transient
  @NlsSafe
  @Attribute("name")
  override var name: String = "",
  @Property(style = Property.Style.ATTRIBUTE, surroundWithTag = false)
  override val server: GiteeServerPath = GiteeServerPath(),
  @Attribute("id")
  @VisibleForTesting
  override val id: String = generateId())
  : ServerAccount() {

  override fun toString(): String = "$server/$name"
}